### PR TITLE
Improve Exception messages

### LINKED ISSUE #485: More detailed Error messages
Could we include some more detailed logging messages to help users debug exceptions? Particularly for `Exception`s caused by `@Injection`s, and also MicroProfile Config 2.0 specific features.

For example, Injecting a `@ConfigProperties` bean where invalidPrefix.akey doesn't exist, e.g.:
```
@Inject
@ConfigProperties(prefix = "invaldPrefix")
ConfigPropertiesPOJO details;
```
Leads to: `NoSuchElementException: SRCFG00014: Property invaldPrefix.akey is required but the value was not found or is empty`

Could this include `@ConfigProperties` specific information? e.g. `No Config Property exists for the prefix <prefix name> for the field <field name>`

Also, I think we should ensure any MicroProfile Config key terms are used and capitalised where appropriate, e.g. the following could be `Required Config Property %s not found`:
https://github.com/smallrye/smallrye-config/blob/bb3d4ab40c720a34b11b6112dabf56690137fe1d/cdi/src/main/java/io/smallrye/config/inject/InjectionMessages.java#L34

### PR DESCRIPTION
**Pull request by Joseph-Cass**
This PR is similar to #513, but doesn't change any validation for `@ConfigMapping` and `@ConfigProperties`.

Resolves #485.

### CODE REVIEW

File: cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java [hunk 1]
```diff
@@ -48,56 +48,37 @@ private ConfigProducerUtil() {
...
+            return SmallRyeConfig.convertValue(name, resolvedValue, resolveConverter(type, config));
         }
 
         BiFunction<Converter<T>, IntFunction<Collection<T>>, Collection<T>> indexedConverter = (itemConverter,
                 collectionFactory) -> {
             Collection<T> collection = collectionFactory.apply(indexedProperties.size());
             for (String indexedProperty : indexedProperties) {
                 // Never null by the rules of converValue
-                collection.add(convertValue(indexedProperty, itemConverter, null, config));
+                collection.add(
```
[the remark this answers is not in the document]
Comment (Joseph-Cass) [reply cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java:80]: Absolutely. Great spot :)

File: cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java [hunk 2]
```diff
@@ -125,6 +106,11 @@ public static String getRawValue(String name, Config config) {
         return SecretKeys.doUnlocked(() -> config.getConfigValue(name).getValue());
     }
 
+    private static String resolveValue(String name, Config config, String defaultValue) {
```
[the remark this answers is not in the document]
Comment (Joseph-Cass) [reply cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java:109]: Okay, this makes sense.

File: cdi/src/main/java/io/smallrye/config/inject/InjectionMessagesUtil.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     * <br>
+     * 
+     * <b>Constructor parameters</b><br>
+     * Given: public
+     * io.smallrye.config.inject.ValidateInjectionTest$ConstructorUnnamedPropertyTest$ConstructorUnnamedPropertyBean(java.lang.String)<br>
+     * Returns:
+     * io.smallrye.config.inject.ValidateInjectionTest$ConstructorUnnamedPropertyTest$ConstructorUnnamedPropertyBean(String)
+     * 
+     */
+    public static String formatInjectionPoint(InjectionPoint injectionPoint) {
```
[the remark this answers is not in the document]
Comment (Joseph-Cass) [reply cdi/src/main/java/io/smallrye/config/inject/InjectionMessagesUtil.java:43]: Can do

File: implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java [hunk 4]
```diff
@@ -224,13 +211,37 @@
...
+     * 2. {@link NoSuchElementException} - if the property is not defined <br>
+     * 3. {@link NoSuchElementException} - if the property is defined as an empty string <br>
+     * 4. {@link NoSuchElementException} - if the {@link Converter} returns {@code null} <br>
+     * <br>
+     * 
+     * Calls for converting optional values ({@link Config#getOptionalValue} and "Injecting Optional Values")
+     * should only throw an {@link Exception} for #1 ({@link IllegalArgumentException} when the property cannot be converted to
+     * the specified type).
+     */
+    public static <T> T convertValue(String name, String value, Converter<T> converter) {
```
[the remark this answers is not in the document]
Comment (Azquelt) [reply implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java:236]: Arguably, this shouldn't be a public method at all. If it wasn't called from `ConfigProducerUtil` it could be private. Would it be better to have this method in a separate class anyway?

If you do want it in the API, then I'm in two minds. At the moment it doesn't use any state at all and it can be quite nice to know that a method operates only on its parameters. It confirms that it really is that simple and it makes it easier to call. On the other hand, if it's behaviour could change and it might need the object state in the future, then it might be better to make it non-static now.

File: cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java [hunk 5]
```diff
@@ -48,56 +48,37 @@ private ConfigProducerUtil() {
...
-        }
-        if (converted == null) {
-            throw InjectionMessages.msg.propertyNotFound(name);
-        }
-        return converted;
-    }
-
-    public static <T> T convertValues(String name, Type type, String defaultValue, Config config) {
+    public static <T> T convertValues(String name, Type type, String resolvedValue, String defaultValue, Config config) {
         String rawValue = getRawValue(name, config);
```
[the remark this answers is not in the document]
Comment (Joseph-Cass) [reply cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java:69]: I see. I think this will need a little shuffle round of methods in `ConfigProducerUtil` due to this edge case in indexedProperties: https://github.com/smallrye/smallrye-config/blob/3e04ed5be90e31fab3a30342d8118b24f6640711/cdi/src/test/java/io/smallrye/config/inject/IndexedPropertiesInjectionTest.java#L96-L98

File: cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Joseph-Cass) [reply cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java:80]: I think https://github.com/smallrye/smallrye-config/blob/fb4e21295a54a7053fcb5e0495da42275abc21da/cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java#L81

Should probably also be 
```
SmallRyeConfig.convertValue(indexedProperty, resolveValue(indexedProperty, config, null), itemConverter));
```
since I don't think an individual `indexedProperty` can ever have a `defaultValue`

[the remark this answers is not in the document]
Comment (radcortez) [reply cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java:81]: Yes.

File: implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (radcortez) [reply implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java:236]: > Arguably, this shouldn't be a public method at all. If it wasn't called from `ConfigProducerUtil` it could be private. Would it be better to have this method in a separate class anyway?
> 

We could move it to another class, but we don't have a suitable one. I think it doesn't fit with `Converters` and I'm trying to keep the class count low, since we are mostly single package :)

> If you do want it in the API, then I'm in two minds. At the moment it doesn't use any state at all and it can be quite nice to know that a method operates only on its parameters. It confirms that it really is that simple and it makes it easier to call. On the other hand, if it's behaviour could change and it might need the object state in the future, then it might be better to make it non-static now.

I agree with your statement, but in this case even if the method only operates in its parameter, we should make it not static, for consistency with the rest of the API, plus we don't really have any other static methods there.

**Comment by radcortez** [comment]
Almost there. Just one comment to address :)

File: implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Joseph-Cass) [reply implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java]: I have made the method non-static

**Review by radcortez — APPROVED** [review]

**Comment by radcortez** [comment]
@Joseph-Cass thanks for the PR and patience :)

**Comment by Joseph-Cass** [comment]
> thanks for the PR and patience
@radcortez No worries :) Thanks for your help.
